widget/list.go

```diff
@@ -0,0 +1,405 @@
+	DisableableWidget
+
+	onTapped          func()
+	statusIndicator   *canvas.Rectangle
+	child             fyne.CanvasObject
+	hovered, selected bool
+}
+
+func newListItem(child fyne.CanvasObject, tapped func()) *listItem {
+	li := &listItem{
+		child:    child,
+		onTapped: tapped,
+	}
+
+	li.ExtendBaseWidget(li)
+	return li
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to its renderer
+func (li *listItem) CreateRenderer() fyne.WidgetRenderer {
+	li.ExtendBaseWidget(li)
+
+	li.statusIndicator = canvas.NewRectangle(theme.BackgroundColor())
+
+	objects := []fyne.CanvasObject{li.statusIndicator, li.child}
+
+	return &listItemRenderer{widget.NewShadowingRenderer(objects, widget.ButtonLevel), li, layout.NewHBoxLayout()}
+}
+
+// MinSize returns the size that this widget should not shrink below
+func (li *listItem) MinSize() fyne.Size {
+	li.ExtendBaseWidget(li)
+	return li.BaseWidget.MinSize()
+}
+
+// MouseIn is called when a desktop pointer enters the widget
+func (li *listItem) MouseIn(*desktop.MouseEvent) {
+	li.hovered = true
+	li.Refresh()
+}
+
+// MouseMoved is called when a desktop pointer hovers over the widget
+func (li *listItem) MouseMoved(*desktop.MouseEvent) {
+}
+
+// MouseOut is called when a desktop pointer exits the widget
+func (li *listItem) MouseOut() {
+	li.hovered = false
+	li.Refresh()
+}
+
+// Tapped is called when a pointer tapped event is captured and triggers any tap handler
+func (li *listItem) Tapped(*fyne.PointEvent) {
+	if li.onTapped != nil && !li.Disabled() {
+		li.selected = true
+		li.Refresh()
+		li.onTapped()
+	}
+}
+
+type listItemRenderer struct {
+	*widget.ShadowingRenderer
+
+	item   *listItem
+	layout fyne.Layout
+}
+
+// MinSize calculates the minimum size of a listItem.
+// This is based on the size of the status indicator and the size of the child object.
+func (li *listItemRenderer) MinSize() (size fyne.Size) {
+	li.item.statusIndicator.SetMinSize(fyne.NewSize(theme.Padding(), li.item.child.Size().Height))
+	size = fyne.NewSize(theme.Padding()+li.item.child.Size().Width, li.item.child.Size().Height)
+	return
+}
+
+// Layout the components of the button widget
+func (li *listItemRenderer) Layout(size fyne.Size) {
+	li.LayoutShadow(size, fyne.NewPos(0, 0))
+	li.layout.Layout([]fyne.CanvasObject{li.item.statusIndicator, li.item.child}, size)
+}
+
+func (li *listItemRenderer) Refresh() {
+	if li.item.selected {
+		li.item.statusIndicator.FillColor = theme.FocusColor()
+	} else if li.item.hovered {
+		li.item.statusIndicator.FillColor = theme.HoverColor()
+	} else {
+		li.item.statusIndicator.FillColor = theme.BackgroundColor()
+	}
+	canvas.Refresh(li.item.super())
+}
+
+type listLayout struct {
+	list      *List
+	itemCount int
+}
+
+var _ fyne.Layout = (*listLayout)(nil)
+
+func newListLayout(list *List) fyne.Layout {
+	return &listLayout{list: list}
+}
+
+func (l *listLayout) Layout(objects []fyne.CanvasObject, size fyne.Size) {
+	if l.list.offsetY != 0 {
```

Comment:
I am curious about this - why return from layout if it's scrolled?

Reply:
Layout does a full layout/restacking of the items.  If it is a scroll we want to leave the items where they are just append/prepend items above or below rather than a full relayout.

Reply:
So we have a special case when this widget is first shown?
Will this function correctly if the offset is non-zero before we first show?

Reply:
No. I dont think it would. Though that's not possible currently and would only become possible if we exposed list's internal scroller where offset could be changed before being shown.

Reply:
With this in mind do we need to add the ScrollToBottom/ScrollToTop that someone recently added to scroller so that we're not stuck with always-top?

Reply:
Good question - How much of scroller functionality do we want to expose?  Changing offsets? Scrolling to bottom/top? Callbacks?

Reply:
From the chat on Slack it seems the agreement is to hold off on exposing scroller functionality for now